### ISSUE TITLE
framerate_range errors with h264 encoder

### ISSUE THREAD

**Issue by RDReavis** [waveform80/picamera#502]
Using PiCamera's framerate_range property results in the following error when attempting to record using the h264 encoder:

```
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    format='h264')
  File "/home/pi/.local/lib/python3.7/site-packages/picamera/camera.py", line 1046, in start_recording
    camera_port, output_port, format, resize, **options)
  File "/home/pi/.local/lib/python3.7/site-packages/picamera/camera.py", line 723, in _get_video_encoder
    self, camera_port, output_port, format, resize, **options)
  File "/home/pi/.local/lib/python3.7/site-packages/picamera/encoders.py", line 601, in __init__
    parent, camera_port, input_port, format, resize, **options)
  File "/home/pi/.local/lib/python3.7/site-packages/picamera/encoders.py", line 187, in __init__
    self._create_encoder(format, **options)
  File "/home/pi/.local/lib/python3.7/site-packages/picamera/encoders.py", line 729, in _create_encoder
    framerate = self.parent.framerate + self.parent.framerate_delta
  File "/home/pi/.local/lib/python3.7/site-packages/picamera/camera.py", line 2409, in _get_framerate_delta
    'framerate_delta cannot be used with framerate_range')
picamera.exc.PiCameraValueError: framerate_delta cannot be used with framerate_range
```

The following code was used to produce the error:
```python
#!/usr/bin/env python3.7

import picamera
import picamera.streams

cam = picamera.PiCamera(
    resolution=(640, 480),
    framerate_range=(1, 15))
stream = picamera.streams.PiCameraCircularIO(
    cam,
    seconds=10,
    bitrate=25000000)
cam.start_recording(
    stream,
    format='h264')
```

Produced using Python 3.7 on Raspberry Pi 3 Model B, picamera 1.13.

**Comment by 6by9** [comment]
What you are trying to do has the potential to work from a firmware level, however:
- Variable frame rate (which is what you're trying to set) is going to cause significant variation in bitrate as the codec has to work from timestamps only, and relies on prediction of how many frames it has to split your 25Mbits over.
- Circular buffers are going to have similar issues due to not knowing how many frames will correspond to 10 seconds. The codec only inserts the header bytes required for splitting the stream every N frames, and N will be a fixed number. The split points are quite likely not be at useful intervals.

What is your use case that requires frame rate variation?

**Comment by RDReavis** [comment]
The use case is to adapt frame rate based on lighting, such that if the exposure duration must be increased due to low light conditions, the frame rate would be decreased to accommodate. I may need to reevaluate my approach if the camera parameter adjustment is best done manually, which would eliminate complexity when it comes to placing the data into a container like mp4 since the frame rate would be known upfront, but would also mean that the camera would need to be stopped and restarted to change the frame rate.

I understand now from your response that there are codec complications arising from this usage. I am less concerned about the circular buffer issues; I sized it according to the codec's maximum bit rate, and decided that the temporal SPS interval variability is currently not critical for my use case.

If `framerate_range` is not to be used in conjunction with `start_recording()`, it would be helpful to have a better error to alert the user that these cannot be used together. The current error encountered gives the impression that `framerate_delta` was being set, which isn't the case:
`picamera.exc.PiCameraValueError: framerate_delta cannot be used with framerate_range`

An update to the docs specifying which recording methods are (un)supported when setting this property would also be helpful. This is an excerpt from the current doc text for `framerate_range` that might be misleading if it is not usable with all recording methods:

> When set, the property configures the camera so that the next call to recording and previewing methods will use the new framerate range.

https://picamera.readthedocs.io/en/release-1.13/api_camera.html#picamera.PiCamera.framerate_range

Thanks for your help!

**Comment by 6by9** [comment]
I maintain the Pi camera firmware, not PiCamera.
The firmware should be capable of doing this, but the way that PiCamera is configuring things during setting up the encoder is throwing the error.
I'll leave it for waveform80 to suss out what the correct behaviour ought to be (I personally loathe Python!)

**Comment by waveform80** [comment]
Doh! Looks like something needs fixing here; PiCamera is attempting to grab `framerate_delta` (in order to calculate the total framerate) when an encoder is initialized, but the getter for `framerate_delta` objects if `framerate_range` is set. I'll get this fixed up for 1.14.
